[WIP] docs: Add design document

orelmisan opened the pull request:
This adds the project's primary design document.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/e016390fa51b57c39cece02a29b2f409435ef3f6..0c76b9b01921dbfac14096d8c12d5dbd14b2c3c1): Added table which compares existing similar projects.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/448a5afbb4d1772d2d318698ceec1b3596e22a1b..bde0a4bf82ca41bd386a034994c447632031965c): Similar projects comparison table - changed a link to a file in [test-network-function](https://github.com/test-network-function/test-network-function) to belong to a specific tag.

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,571 @@
+
+#### Overview
+
+**Note**: The framework uses well-defined [ConfigMap](#configuration-and-output-specification) and [Job](#job-specification) objects (described below).
```
Loved it, thanks.

orelmisan commented on docs/kiagnose_design.md:
Done.

orelmisan commented on docs/kiagnose_design.md:
Done.

orelmisan replied:
`## Permission Management` is now right under `# Appendix C - Design Decisions Considerations`.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/bde0a4bf82ca41bd386a034994c447632031965c..2743b8ef3d347711688fac0758cc7b72de1dfb76): Fixed code review comments.

EdDev commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,588 @@
+- Collection of tests that are grouped together under the same project with basic tooling to execute them.
+- Collection of tests that are grouped together for certification with tooling that help integrate their run and result posting to Red Hat ([preflight](https://github.com/redhat-openshift-ecosystem/openshift-preflight)).
+- Framework which can run arbitrary workloads ([Sonobuoy](https://sonobuoy.io/)).
+  The first two groups focus on self curated tests and not on arbitrary ones.
```
Newline is needed above.

orelmisan replied:
Done.

EdDev commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,588 @@
+These solutions may be grouped into the following categories:
+
+- Collection of tests that are grouped together under the same project with basic tooling to execute them.
+- Collection of tests that are grouped together for certification with tooling that help integrate their run and result posting to Red Hat ([preflight](https://github.com/redhat-openshift-ecosystem/openshift-preflight)).
```
This line is too long. I would try to wrap lines to some reasonable length.
72, 100, whatever looks nice.

EdDev commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,588 @@
+
+##### Configuration Fields
+
+| Key                                  | Description                                                                              | Is Mandatory | Remarks                                                                      |
+|--------------------------------------|------------------------------------------------------------------------------------------|--------------|------------------------------------------------------------------------------|
+| `spec.image`                         | Checkup image                                                                            | True         |                                                                              |
+| `spec.timeout`                       | Timeout to wait for a checkup to finish [min]                                            | True         |                                                                              |
+| `spec.param.<checkup argument name>` | Arbitrary parameter required by the concrete checkup                                     | False        | 0..N                                                                         |
+| `spec.clusterRoles`                  | ClusterRole required by the checkup Value format: `<ClusterRole name>` Newline separated | False        | 0..N e.g: `default/nad-reader` Should already exist before launching the Job |
+| `spec.roles`                         | Role required by the checkup Value format: `<namespace/name>` Newline separated          | False        | 0..N e.g: `default/nad-reader` Should already exist before launching the Job |
```
Can some columns be shrink by splitting the text with a newline?
It may look fine on a rendered page, but to review this it is hard.

EdDev reviewed (commented):
Thank you, looks great.Just a few inline comments.

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,588 @@
+2. The user shall create a Job object in order to run the framework.
+3. The user shall wait for the checkup Job to finish and get the checkup results from the ConfigMap object.
+
+> **_NOTE:_** This interface could evolve in the future to a CRD and a dedicated controller.
```
Done.

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,588 @@
+- Provide input for the selected checkup.
+- Receive an output at the end of the checkup run.
+
+> **_NOTE:_** It shall be the user’s responsibility to remove the ConfigMap object after the results are read.
```
We have it as a note under "Checkup Results Removal" section.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/2743b8ef3d347711688fac0758cc7b72de1dfb76..2262db5dd83cef162778896029fda00d0a8a6b17): Fixed review comments.

ormergi commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+## Exploration of Similar Projects
+
+Exploring existing solutions revealed several projects which aimed to run various tests on running clusters.
+These solutions may be grouped into the following categories:
```
This line wont start at new line.

orelmisan replied:
Done.

ormergi commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+## Exploration of Similar Projects
+
+Exploring existing solutions revealed several projects which aimed to run various tests on running clusters.
+These solutions may be grouped into the following categories:
```
This line wont start at new line.

orelmisan replied:
Is this comment a duplicate of the previous comment?

ormergi commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+result posting to Red Hat ([preflight](https://github.com/redhat-openshift-ecosystem/openshift-preflight)).
+- Framework which can run arbitrary workloads ([Sonobuoy](https://sonobuoy.io/)).
+
+The first two groups focus on self curated tests and not on arbitrary ones.
```
Typo  "The first two groups focus on self curated tests and not **on** arbitrary ones.", it should be "an".

orelmisan replied:
This is not a typo, I've verified it with a [grammar checking tool](https://languagetool.org/).

ormergi commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+result posting to Red Hat ([preflight](https://github.com/redhat-openshift-ecosystem/openshift-preflight)).
+- Framework which can run arbitrary workloads ([Sonobuoy](https://sonobuoy.io/)).
+
+The first two groups focus on self curated tests and not on arbitrary ones.
```
Typo  "The first two groups focus on self curated tests and not **on** arbitrary ones.", it should be "an".

orelmisan replied:
Is this comment a duplicate of the previous one?

ormergi reviewed (commented):
Hi, it looks good, please see inline comments

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+  namespace: $RESULT_CONFIGMAP_NAMESPACE
+data:
+  status.succeeded: "true"
+  status.failureReason: none
```
Done.

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+  spec.timeout: 1m
+  spec.param.message: "Hi!"
+  status.succeeded: “true”
+  status.failureReason: none
```
Done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/2262db5dd83cef162778896029fda00d0a8a6b17..0d53ad02fd95fbcd5092970a1ee6839619e641a7): Fixed review comments by @ormergi.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+Kiagnose - Kubernetes Checkup Framework
```
Regarding the diagram, it could be even cooler if it was in svg (which smaller to save and easier to modify)

EdDev replied:
@orelmisan , could you please respond?

orelmisan replied:
Done.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+Kiagnose - Kubernetes Checkup Framework
+=
+
+**Authors**: Edward Haas, Orel Misan, Or Mergi
```
care to add your @orelmisan for "bragging rights"?

orelmisan replied:
Done.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+
+# Summary
+
+This proposal introduces a framework for running K8s cluster checkups, which verifies that a certain cluster
+functionality is working as expected.
```
```suggestion
This proposal introduces a framework to verify that certain cluster
functionalities are working as expected.
```

I think that at this stage, the reader does not yet know what is a "cluster checkup", so a shorter summary may be better.

orelmisan replied:
Done.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+
+# Motivation
+
+Operating a K8s cluster is challenging. Such clusters usually provide a wide range of functionalities and services, some
```
```suggestion
Operating a [K8s](https://kubernetes.io/) cluster is challenging. Such clusters usually provide a wide range of functionalities and services, some
```
in case your reader does not know what k8s is.

orelmisan replied:
Done.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+
+## Goals
+
+- Expose a checkup to a cluster.
```
I don't understand this goal

EdDev replied:
The cluster needs to know the checkup is there.
At the minimum, we need to tell someone where it is and it should be accessible from the cluster.

Originally we thought about it as "register checkup to the cluster" but there was some resistance and it was too specific.

dankenigsberg replied:
I still do not understand it... The checkup framework that you suggest here is responsible to expose *a* checkup? Or is it about exposing multiple checkups, some of them predefined, some of them user-defined, some of them by third parties?

orelmisan replied:
@EdDev @dankenigsberg WDYT about "Deploy a checkup to a cluster"?

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+  termination.
+- Checkup execution should require minimal dependencies, allowing a cluster administrator or some management tool to
+  trigger it.
+- Deploy and run the checkups on any K8s based cluster.
```
```suggestion
- Deploy and run the checkups on any K8s-based cluster.
```

orelmisan replied:
Done.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+- Checkup execution should require minimal dependencies, allowing a cluster administrator or some management tool to
+  trigger it.
+- Deploy and run the checkups on any K8s based cluster.
+- Provide a well-defined API, allowing management components to interact with the solution.
```
```suggestion
- Provide a well-defined API, allowing management components to interact with the framwork.
```

I prefer using the same term for what you build here. (unless you mean something else by "solution")

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+
+- Defining content of the checkup, including its correctness, quality and reliability, is up to its author.
+- GUI is not in the scope of this document.
+- Deploy clusters and components that a predefined checkup may require.
```
I don't understand the English here.

EdDev replied:
It tries to say that dependencies the checkup requires are not going to be deployed or managed by kiagnose.

EdDev replied:
@orelmisan , would you want to try and rephrase it?

orelmisan replied:
@EdDev @dankenigsberg WDYT about "Deploying checkup's dependencies"?

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+
+# Proposal
+
+## Definition of Users
```
```suggestion
## Definition of Personas
```
I find it as a clearer term.

orelmisan replied:
Done.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,597 @@
+
+- K8s cluster administrators.
+- Vendors who provide verification of their K8s extensions.
+- Installation, integration & support teams, which examine cluster functionality.
```
How is installation team different from cluster admin?
Can you focus the third persona on a single support engineer? I find it confusing to consider three types of teams as a persona.

EdDev replied:
> How is installation team different from cluster admin?

One installs the cluster, possibly for the first time.
While the Admin may just maintain it afterwards.

dankenigsberg replied:
I don't see how the differentiation between the two helps the discussion here. Also I don't understand how the installation team is related to the proposed framework, and why they should be bundled with the support team. I still think that the persona definition would be improved if you focus on one of them.

orelmisan replied:
The intent was "people which installs a cluster and people which support existing clusters".

dankenigsberg replied:
I don't see how "people which install a cluster" is an interesting persona here, nor how its behavior and requirement is different from that of the cluster admin.

orelmisan replied:
We can change these personas to:
- cluster admins (and all the other personas with similar permissions).
- namespace admins.
- vendors.

WDYT?

dankenigsberg reviewed (changes requested):
Thanks for this proposal! Sorry that I'm providing here only a very limited review.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/0d53ad02fd95fbcd5092970a1ee6839619e641a7..002e2fb655a61bfc4aadcfa3f9f7c62e6493cd4c): Fixed review comments suggested by @dankenigsberg.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+
+## User Stories
+
+As a cluster administrator:
```
Can we somehow refine this? I can think of another a non-admin, who got access to a cluster and now would like to check if the cluster fits their needs.

I think that it is important that a non-privileged user would be able to run the checkups. If a specific checkup fails due to lack of permissions, that *exactly* what the user needs to know.

orelmisan replied:
@dankenigsberg, we consider a checkup to be an untrusted arbitrary code running inside a cluster.
The framework can bind existing ClusterRole objects with the checkup's ServiceAccount - the user only have provide a ClusterRole name and it is automatically bound.
In other words, if you know a ClusterRole's name - you can use it.

From a security standpoint, IMO the users that could configure and run checkups should be trusted, because they wield a great power.

In order to configure and run a checkup, a user is required to have the following permissions in the framework's namespace:
1. Read/write ConfigMaps (or a CR in the near future).
2. Create a Job.

So full admin privileges are not required in order to use the framework.

dankenigsberg replied:
I think that the use case of non-privileged tester should be written in this document even if the current implementation does not support it. If I was a cluster admin, I would *not* run some random checkup container as an admin. I'd prefer to have a clear  request on which namespace I need to create, which permissions should I grant to the namespace admin, and only then run the checkup.

EdDev replied:
This was not a goal or requirement.
Kiagnose is privileged and trusted, with elevated permissions (e.g. create namespaces for the checkup workload).
On the other hand, the checkup themselves are untrusted and requires the Admin to register their permissions explicitly before usage.

When (and if) kiagnose will start using a controller with a CR, we would probably be able to give permissions to who can create such CR/s and by that allow different users to execute checkups.

EdDev replied:
@dankenigsberg , would you like to add to this thread?
We would like to merge this design proposal before the first release tag, so I would like to make sure we also finalize this topic.

dankenigsberg replied:
> would you like to add to this thread?

Sorry, I missed your month-old questions.

> This was not a goal or requirement.

That's why a proposal review is important.

I have an OpenShift cluster with a very strict admin. Only he can create projects (=namespaces) and set their admin users. I can ask him to grant special rights to a project of mine (e.g more quota - the default is very very small, assign a specific NAD to it etc) but he would outright reject installing an opensource deployment with unlimited permissions.

He would be fine if I run the checkup on *my* project, consuming *my* quota, testing the specific NAD allocated to me, under *my* privileges.

I'm not asking to do it *now*, but to me, it is important that the framework has a lightweight mode which expects all resources to be prepared ahead of time. I don't mind if it also has another module that attempts to create the test namespace and clean it after the fact.

dankenigsberg replied:
> I agree that the framework has a lot of power, which could lead to security problems.

It's not just that. The cluster I talk about has annoyingly low default quota. Unless the framework can outguess the quota required by the individual checkup, it would not run properly. And if it does whatever the checkup declare it needs, it would be easy for a malicious checkup to starve the system.

> I think we should discuss what are the requirements and limitations when running a checkup in an existing namespace.

+1

> One obvious limitation will probably be that the framework could not enforce a proper cleanup of all objects created by a checkup.

I'm ok with documenting that a  buggy/malicious checkup may leave content in my namespace, which I'm advised to empty after usage.

> a checkup (or the framework) could not run under a user's privileges

I can run a pod and a VM under my human privileges, why not a checkup?

If I must make the service account an admin of my namespace before running the checkup, so be it. My cluster-admin would not be happy if this account is too privileged, though.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/002e2fb655a61bfc4aadcfa3f9f7c62e6493cd4c..92b461b97662dcc8f3b6a85c1f90a8c44da72777): Rebase
Changed block diagram format to `svg`.

RamLavi reviewed (changes requested):
The design document is fairly clear.

However I find talking about the two separate configmaps confusing. Please go over the design doc and fill in specific names where we are talking about the user-supplied cm and when the result cm. This should also be stated on the diagram.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/5390e5c06a7e1ec8c19a68cf9abd3dd819b03529..3d937d4513d60a619a82970e0b38a22ce2316ad6): Rebase
Update block diagram.

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+
+##### Output Fields
+
+1. The user shall expect the checkup results concatenated to the ConfigMap object at the end of the checkup run.
```
Here it is actually the user-supplied ConfigMap object.
I've made it explicit.

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,596 @@
+> This interface could evolve in the future to a 
+> [CRD](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) and a dedicated controller.
+
+1. The user shall create a **mutable** ConfigMap object in the `k8s-checkup-framework` namespace in order to:
```
Done all over.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,571 @@
+
+- Defining content of the checkup, including its correctness, quality and reliability, is up to its author.
+- GUI is not in the scope of this document.
+- Deploy clusters and components that a predefined checkup may require.
```
```suggestion
- Deploying clusters
- Deploying components that a specific checkup may require.
```

orelmisan replied:
Done.

RamLavi reviewed (approved):
nice! 
/lgtm

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/ae9ab4015ee86ceba764f27a57c2bf7b1255bb1c..6a78ec77fc167f7cc7de17864c825803d107efd9): Answered a [comment](https://github.com/kiagnose/kiagnose/pull/11#discussion_r897549633).

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/6a78ec77fc167f7cc7de17864c825803d107efd9..a224bf758662124bee79d6e91847b444462fea71): Rebase
Fixed heading links.
Added namespace administrator persona.
Added namespace administrator use case to third implementation phase.

dankenigsberg commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,582 @@
+    - I can export the checkup results, cluster state and all other relevant information, so I can analyze them or pass
+      them on to experts.
+
+As a namespace administrator:
```
thanks, /lgtm

ormergi reviewed (commented):
Overall looks good:
- Updating the image tags to the ones available at `quay.io/kiagnose` will be nice.

- Please consider change the wording "user-supplied ConfigMap" to "checkup ConfigMap".I dont think user-supplied is the right wording is doesn't sound right to me.

Also using "installation" doesnt sound right (to me), we should use "deployment / deploying" terminology.



The notes sections are not rendered properly, please make sure each line starts with `>`, it should fix it.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/9cd2d89fae50fb6b89f5f465564c76166dc34ef1..1b20783a5017d8c97760a9fe3c448fd3ff42b556): Fixed notes rendering.

orelmisan commented:
@ormergi 
> * Updating the image tags to the ones available at `quay.io/kiagnose` will be nice.

Done.

> * Please consider change the wording "user-supplied ConfigMap" to "checkup ConfigMap".
>   I dont think user-supplied is the right wording is doesn't sound right to me.

It is supplied by the user, and serves as our main user-interface.
The term `user-supplied ConfigMap` appears in all of our documentation, feel free to suggest a change in a follow-up PR.

>   Also using "installation" doesn't sound right (to me), I think we should use "deployment / deploying" instead.

The term `installation` appears in all of our documentation, feel free to suggest a change in a follow-up PR.

> * The notes sections are not rendered properly, please make sure each line starts with `>` or `> **_Note_**:` (no caps, punctuation at the end), it should fix it.

Thanks, done.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/1b20783a5017d8c97760a9fe3c448fd3ff42b556..2213268734f4a2f3437fd66e4d6d66e5a7ff2125): Fixed comments made by @EdDev in an offline discussion.

ormergi commented:
> It is supplied by the user, and serves as our main user-interface. The term `user-supplied ConfigMap` appears in all of our documentation, feel free to suggest a change in a follow-up PR.

Using supply is wrong, you should use "supply" to describe some kind of resource that one party provide for another.
As far as I understand a checkup configuration is not a resource.

> The term `installation` appears in all of our documentation, feel free to suggest a change in a follow-up PR.

Its a shame that I didn't catch these earlier,I could at least try to prevent other docs to use this wording (taking about both)
If you do agree you can do those changes to this doc (and also at #12), then the follow up PR will be just for
changing the docs that already exists.

ormergi reviewed (approved):
I am ok with merging this PR the way it is, I will open follow up PR for the terminology begin used all over the docs.

EdDev commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,579 @@
+
+### Checkup Setup
+#### Prerequisites
+1. Checkup Role and ClusterRole objects exist in the cluster.
```
Odd to see a single number.

EdDev commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,579 @@
+### Review Checkup Results
+#### Prerequisites
+The cluster user / ServiceAccount should have the permissions to list and read the following objects:
+1. A user-supplied ConfigMap object in the `kiagnose` namespace.
```
Odd to see a single number.

orelmisan replied:
Done.

EdDev commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,579 @@
+1. A user-supplied ConfigMap object in the `kiagnose` namespace.
+
+#### Process Description
+1. The user shall read the user-supplied ConfigMap object they had previously created.
```
Odd to see a single number.

orelmisan replied:
Done.

orelmisan commented on docs/kiagnose_design.md:
Done.

orelmisan commented on docs/kiagnose_design.md:
```diff
@@ -0,0 +1,579 @@
+### Checkup Results Removal
+#### Prerequisites
+The cluster user / ServiceAccount should have the permissions to list and read and delete the following objects:
+1. A user-supplied ConfigMap object in the `kiagnose` namespace.
```
Done.

EdDev reviewed (approved):
Thank you!
